### DISCUSSION TITLE
Support non-systemd environments

### DISCUSSION THREAD

**Discussion by ghost** [mullvad/mullvadvpn-app#9054] [Feature requests & Ideas]
## Issue report
Operating system: Gnuinos (Based Devuan) 
App version: 2023.4

### Issue description
/var/lib/dpkg/info/mullvad-vpn.postinst:` row 6: systemctl: command not found
dpkg: An error occurred while processing package mullvad-vpn (--configure): installed mullvad-vpn package post-installation script subprocess returned error exit status 127
An error occurred during processing: mullvad-vpn
E: Sub-process /usr/bin/dpkg returned an error code (1)

In an environment without systemd, you can't use systemctl, so you can't install packages.

**Comment by faern** [comment]
That is correct. This app requires systemd. At least for now. We have not spent any time on supporting anything else since there is so little demand.

**Comment by AllergicMushroom** [comment]
Actually, I've been waiting for that for about 3 months as well. My system runs sysvinit instead of systemd.

Edit: Other possibilities of init systems in my machine are OpenRC and runit.

**Comment by lupin3rd** [comment]
I also use Devuan. There is a way to use mullvad in my system?

**Comment by ghost** [comment]
@lupin3rd Need to build for non-systemd environments.
However, this is a step beyond power users and should be done by developers or hackers.

**Comment by msamesch** [comment]
Maybe installing systemctl and starting the daemon with
`systemctl start mullvad-daemon`
is a workaround.

At least I could start the daemon and the app  on Devuan Daedalus
(`/opt/Mullvad\ VPN/mullvad-gui`).

Now I need to buy credit to test it.

**Comment by ghost** [comment]
> Maybe installing systemctl and starting the daemon with
> systemctl start mullvad-daemon
> is a workaround.

@msamesch This report is for OpenRC and runit users who don't plan to use systemd.

> At least I could start the daemon and the app on Devuan Daedalus
> (/opt/Mullvad\ VPN/mullvad-gui).

I couldn't get the GUI to open at all.

**Comment by msamesch** [comment]
> @msamesch This report is for OpenRC and runit users who don't plan to use systemd.

@Noir16 
> > At least I could start the daemon and the app on Devuan Daedalus

so no systemd either. I'm using SysVinit.

When installing mullvad-vpn together with systemctl (`Conflicts: systemd`, `Description: daemonless "systemctl" command to manage services without systemd`), it still complains:

`ERROR:systemctl:Unit /usr/lib/systemd/system/mullvad-daemon.service could not be found.`

When I replace the referring lines in /var/lib/dpkg/info/mullvad-vpn.postinst with

`systemctl enable mullvad-daemon.service` and
`systemctl enable mullvad-early-boot-blocking.service` respectively, it installs without an error.

**Comment by SlySven** [comment]
I was looking for a VPN for various OSes (Windows, GNU/Linux {Devuan 5.0 "daedalus"} and FreeBSD) and this looked the most promising - until I saw the supposed hard dependency on `systemd`. *\*sigh\**.

**Comment by AllergicMushroom** [comment]
Not sure if this helps you, but I managed to have it run on Artix Linux which has no systemd by installing mullvad-vpn and mullvad-vpn-openrc.

**Comment by sys-fs** [comment]
Another Devuan user here. Forgive me if dropping a link to my own repository is frowned upon, but I've written [a script](https://github.com/sys-fs/mullvad-repack) that patches the Mullvad client's Debian package with support for sysv, openrc, or runit depending on which init system it detects. Hopefully it's helpful if any of you are still using Devuan, or if anyone else ends up on this discussion thread that does.

**Reply by sys-fs** [reply]
The repo now has an apt repo published from it, so you don't need to patch the package locally anymore.

**Comment by Moonman0922** [comment]
I think this is more relevant than ever with the news that systemd is implementing age verification. Has there been progress on this lately?

**Reply by AllergicMushroom** [reply]
I don't think devs will consider this more relevant with age verification laws.

That said, at some point there was a mullvad-openrc package in AUR that I used when I was running Artix. It actually ran better than the Arch package (also non official, since Mullvad only supports Ubuntu, Debian, and Fedora).

Note also that you can use Wireguard with Mullvad, it's a solid workaround.

**Reply by arc-d3v** [reply]
There are still some packages that add services for different sys inits in AUR.
mullvad-vpn-openrc
mullvad-vpn-dinit
...
You can use these or even create your own service using these packages as example.

**Comment by katywings** [comment]
Another devuan user here, @sys-fs's script from https://github.com/sys-fs/mullvad-repack was a life saver ':)

**Reply by sys-fs** [reply]
Glad it's useful for you. I've set up an apt repo from that git repo now, so you won't have to patch the package every time you want to upgrade anymore. :-)
